File: GanttChartLightLibraryDemos/Demos/Samples.Resources/WPF-CSharp/GanttChartDataGrid/Database/PredecessorItemsConverter.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Windows.Data;
using System.Globalization;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using DlhSoft.Windows.Controls;
using System.ComponentModel;
using System.Linq;
using System.Collections.Specialized;

namespace GanttChartDataGridDatabaseSample
{
    public class PredecessorItemsConverter : IValueConverter
    {
        // Retrieve a PredecessorItemCollection based on Predecessor data context.
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            IEnumerable<Predecessor> predecessors = (value as IEnumerable<Predecessor>).OrderBy(p => p.Task.Index);
            PredecessorItemCollection items = new PredecessorItemCollection();
            foreach (Predecessor predecessor in predecessors)
            {
                PredecessorItem item = new PredecessorItem { Tag = predecessor };
                SetPredecessorItem(item);
                items.Add(item);
            }
            items.CollectionChanged += Items_CollectionChanged;
            return items;
        }

        // When the PredecessorItem collection changes, update data context accordingly.
        private void Items_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
        {
            switch (e.Action)
            {
                case NotifyCollectionChangedAction.Add:
                    IEnumerable<PredecessorItem> items = sender as IEnumerable<PredecessorItem>;
                    foreach (PredecessorItem item in e.NewItems)
                    {
                        Task task = item.Item.Tag as Task;
                        Predecessor predecessor = new Predecessor { Task = task, DependencyType = (int)DependencyType.FinishStart };
                        item.Tag = predecessor;
                        SetPredecessorItem(item);
                        ContextTask.Predecessors.Add(predecessor);
                    }
                    break;
                case NotifyCollectionChangedAction.Remove:
                    foreach (PredecessorItem item in e.OldItems)
                    {
                        Predecessor predecessor = item.Tag as Predecessor;
                        ContextTask.Predecessors.Remove(predecessor);
                    }
                    break;
            }
        }

        private void SetPredecessorItem(PredecessorItem item)
        {
            Predecessor predecessor = item.Tag as Predecessor;
            BindingOperations.SetBinding(item, PredecessorItem.ItemProperty, new Binding("Task") { Source = predecessor, Mode = BindingMode.TwoWay, Converter = TaskItemConverter.GetInstance(GanttChartItems) });
            BindingOperations.SetBinding(item, PredecessorItem.DependencyTypeProperty, new Binding("DependencyType") { Source = predecessor, Mode = BindingMode.TwoWay, Converter = NumericDependencyTypeConverter.Instance });
        }


        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotSupportedException();
        }

        public Task ContextTask { get; private set; }
        public IEnumerable<GanttChartItem> GanttChartItems { get; private set; }

        public static PredecessorItemsConverter GetInstance(Task contextTask, IEnumerable<GanttChartItem> ganttChartItems)
        {
            return new PredecessorItemsConverter { ContextTask = contextTask, GanttChartItems = ganttChartItems };
        }
    }
}

File: GanttChartLightLibraryDemos/Demos/Samples.Resources/WPF-VisualBasic/GanttChartDataGrid/Database/PredecessorItemsConverter.vb
Imports System.Net
Imports System.Windows.Ink
Imports System.Windows.Media.Animation
Imports System.Globalization
Imports System.Collections.ObjectModel
Imports DlhSoft.Windows.Controls
Imports System.ComponentModel
Imports System.Collections.Specialized

Public Class PredecessorItemsConverter
    Implements IValueConverter

    ' Retrieve a PredecessorItemCollection based on Predecessor data context.
    Public Function Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
        Dim predecessors As IEnumerable(Of Predecessor) = (TryCast(value, IEnumerable(Of Predecessor))).OrderBy(Function(p) p.Task.Index)
        Dim items As New PredecessorItemCollection()
        For Each pred As Predecessor In predecessors
            Dim item As PredecessorItem = New PredecessorItem With {.Tag = pred}
            SetPredecessorItem(item)
            items.Add(item)
        Next pred
        AddHandler items.CollectionChanged, AddressOf Items_CollectionChanged
        Return items
    End Function

    ' When the PredecessorItem collection changes, update data context accordingly.
    Private Sub Items_CollectionChanged(sender As Object, e As NotifyCollectionChangedEventArgs)
        Select Case e.Action
            Case NotifyCollectionChangedAction.Add
                Dim items As IEnumerable(Of PredecessorItem) = TryCast(sender, IEnumerable(Of PredecessorItem))
                For Each item As PredecessorItem In e.NewItems
                    Dim t As Task = TryCast(item.Item.Tag, Task)
                    Dim pred As Predecessor = New Predecessor With {.Task = t, .DependencyType = CInt(Fix(DependencyType.FinishStart))}
                    item.Tag = pred
                    SetPredecessorItem(item)
                    ContextTask.Predecessors.Add(pred)
                Next item
            Case NotifyCollectionChangedAction.Remove
                For Each item As PredecessorItem In e.OldItems
                    Dim pred As Predecessor = TryCast(item.Tag, Predecessor)
                    ContextTask.Predecessors.Remove(pred)
                Next item
        End Select
    End Sub

    Private Sub SetPredecessorItem(item As PredecessorItem)
        Dim pred As Predecessor = TryCast(item.Tag, Predecessor)
        BindingOperations.SetBinding(item, PredecessorItem.ItemProperty, New Binding("Task") With {.Source = pred, .Mode = BindingMode.TwoWay, .Converter = TaskItemConverter.GetInstance(GanttChartItems)})
        BindingOperations.SetBinding(item, PredecessorItem.DependencyTypeProperty, New Binding("DependencyType") With {.Source = pred, .Mode = BindingMode.TwoWay, .Converter = NumericDependencyTypeConverter.Instance})
    End Sub


    Public Function ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
        Throw New NotSupportedException()
    End Function

    Private privateContextTask As Task
    Public Property ContextTask() As Task
        Get
            Return privateContextTask
        End Get
        Private Set(ByVal value As Task)
            privateContextTask = value
        End Set
    End Property
    Private privateGanttChartItems As IEnumerable(Of GanttChartItem)
    Public Property GanttChartItems() As IEnumerable(Of GanttChartItem)
        Get
            Return privateGanttChartItems
        End Get
        Private Set(ByVal value As IEnumerable(Of GanttChartItem))
            privateGanttChartItems = value
        End Set
    End Property

    Public Shared Function GetInstance(contextTask As Task, ganttChartItems As IEnumerable(Of GanttChartItem)) As PredecessorItemsConverter
        Return New PredecessorItemsConverter With {.ContextTask = contextTask, .GanttChartItems = ganttChartItems}
    End Function
End Class
